openedx_tagging/core/tagging/api.py

```diff
@@ -107,3 +107,73 @@ def tag_object(
     """
 
     return taxonomy.tag_object(tags, object_id, object_type)
+
+
+def autocomplete_tags(taxonomy: Taxonomy, prefix: str, object_id: str= None, count=10) -> List[TagResult]:
+    """
+    Returns the first `count` tag values in the given Taxonomy with names 
+    that begin with the given prefix string. The result is returned in alphabetical
+
+    Closed taxonomies return tag values that exist in the taxonomy. Also excludes all Tags that the
+    `object_id` already has. 
+
+    Free-text taxonomies return only tag values that are currently in use on (another) object. Also excludes
+    the tags used by `object_id`.
+    """
+    result = []
+
+    # Fetch tags that the object already has to exclude them from the result
+    excluded_tags = ObjectTag.objects.filter(object_id=object_id, taxonomy=taxonomy)
+    if taxonomy.allow_free_text:
+        # Free-text taxonomy
+
+        # Obtain the value of the excluded tags
+        excluded_tags = excluded_tags.values_list('_value', flat=True)
+        result = (
+            # Fetch object tags from this taxonomy whose value starts with the given prefix
+            ObjectTag.objects.filter(taxonomy=taxonomy, _value__istartswith=prefix)
+            # omit any tags applied to the given object
+            .exclude(object_id=object_id)
+            # omit any free-text tags from other objects whose values match the tags on the given object
+            .exclude(_value__in=excluded_tags)
+            # alphabetical ordering
+            .order_by('_value')
+            # obtain the values of the tags
+            .values_list('_value', flat=True)
+            # remove repeats
+            .distinct()
+            # get only first `count` values
+            [:count]
+        )
+        # Build tag results
+        result = [TagResult(id=None, value=value) for value in result]
+    else:
+        # Closed taxonomy
+        
+        # Obtain the id of the excluded tags
+        excluded_tags = excluded_tags.filter(tag__isnull=False).values_list('tag__id', flat=True)
+        result = (
+            # Fetch tags from this taxonomy whose value starts with the given prefix
+            Tag.objects.filter(taxonomy=taxonomy, value__istartswith=prefix)
+            # omit any tags applied to the given object
+            .exclude(id__in=excluded_tags)
+            # alphabetical ordering
+            .order_by('value')
+            # obtain only id and values of the tags
+            .values('id', 'value')
+            # get only first `count` values
+            [:count]
+        )
+
+        # We need to delete the repeats here because the current database backend 
+        # don't support `distinct()` on fields
+        unique_results = {}
+        for item in result:
+            value = item['value']
+            if value not in unique_results:
+                unique_results[value] = item
+        result = list(unique_results.values())
```

Comment:
Hmm.. I think we should either:

* enforce `unique_together = ("taxonomy", "value")` on the Tag model, OR
* return all the matching Tags, even if there are duplicates.

Because either we won't have multiple tags in a taxonomy with the same name, or we'll confuse the user by only returning the first one available when they're searching for a matching tag.

Sounds like a product question, though I'm hard-pressed to come up with a use case for duplicate tags at the leaf level anyway..

Reply:
> Sounds like a product question, though I'm hard-pressed to come up with a use case for duplicate tags at the leaf level anyway.

It's possible to have two tags with the same value but with different parent? This is the only use case I can think of.
If that can be done, then we can't use `unique_together = ("taxonomy", "value", "parent")`, bacuse doesn't work with foreing keys.
 
I think for now it's ok to return all matching Tags

Reply:
@ChrisChV 

> It's possible to have two tags with the same value but with different parent?

Yep, that's possible from a data perspective, but if Product doesn't require it, then we don't have to make room for it in our implementation.

Can you raise the question with Product, and see if they can find out whether any of the taxonomies they want to use have any duplicate leafs?

Reply:
@pomegranited [Question in Slack](https://openedx.slack.com/archives/C03T0RZEASG/p1688402959991889). As @ormsbee commented, I added the constraint: adb3e5afdba90aec3b78a0ec609cd9fd29b093bb